File: python/49.py
import collections
from typing import List


class Solution(object):
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = collections.defaultdict(list)

        for word in strs:
            anagrams[''.join(sorted(word))].append(word)

        return list[anagrams.values()]


if __name__ == "__main__":
    solution = Solution()

    inputs = ["eat", "tea", "tan", "ate", "nat", "bat"]

    print(solution.groupAnagrams(inputs))

File: cpp/leetcode/49.cpp
#include <vector>
#include <iostream>
#include <unordered_map>
#include <queue>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> map;
        for (auto str : strs) {
            priority_queue<char> order;
            for (auto ch : str) {
                order.push(ch);
            }
            string word;

            while(order.size() > 0) {
                char ch = order.top();
                order.pop();
                word.push_back(ch);
            }

            map[word].push_back(str);
        }

        vector<vector<string>> retVal;
        for (auto item : map) {
            retVal.push_back(item.second);
        }

        return retVal;
    }
};

int main() {
    vector<string> strs = {"eat","tea","tan","ate","nat","bat"};
    Solution solution;
    auto ret = solution.groupAnagrams(strs);

    return 0;
}
